terraform/main.tf

```diff
@@ -0,0 +1,123 @@
+variable "public_key_path" {
+  type = string
+  default = "~/.ssh/id_ed25519.pub"
+  description = "Path to public key for creating a key pair"
+}
+
+provider "aws" {
+  region = "us-east-2"
+}
+
+module "vpc" {
+  source = "terraform-aws-modules/vpc/aws"
+
+  name = "envoy-ci"
+  cidr = "10.0.0.0/16"
+
+  azs             = ["us-east-2a"]
+  private_subnets = []
+  public_subnets  = ["10.0.101.0/24"]
+}
+
+data "aws_ssm_parameter" "debian" {
+  name = "/aws/service/debian/release/11/latest/amd64"
+}
+
+module "security_group" {
+  source = "terraform-aws-modules/security-group/aws//modules/ssh"
+
+  name        = "envoy-ci-ssh"
+  vpc_id      = module.vpc.vpc_id
+
+  ingress_cidr_blocks = ["0.0.0.0/0"]
+}
+
+resource "aws_key_pair" "ci" {
+  key_name = "envoy-ci"
+  public_key = trimspace(file(var.public_key_path))
+}
+
+resource "aws_instance" "envoy-ci-build" {
+  ami           = data.aws_ssm_parameter.debian.value
+  instance_type = "t3.2xlarge"
+
+  iam_instance_profile = aws_iam_instance_profile.envoy-ci-build.name
+
+  key_name = aws_key_pair.ci.id
+
+  tags = {
+    Name = "envoy-ci"
+  }
+
+  root_block_device {
+    volume_size = "50"
+  }
+
+  subnet_id                   = module.vpc.public_subnets[0]
+  vpc_security_group_ids      = [module.security_group.security_group_id]
+
+  user_data = <<EOF
+#!/bin/bash
+set -e
+
+apt-get update
+apt-get install -y \
+    ca-certificates \
+    curl \
+    gnupg \
+    lsb-release \
+    git \
+    make \
+    golang \
+    gpg \
+    python3 \
+    python-is-python3
+
+sudo mkdir -m 0755 -p /etc/apt/keyrings
+curl -fsSL https://download.docker.com/linux/debian/gpg | sudo gpg --dearmor -o /etc/apt/keyrings/docker.gpg
+
+echo \
+  "deb [arch=$(dpkg --print-architecture) signed-by=/etc/apt/keyrings/docker.gpg] https://download.docker.com/linux/debian \
+  $(lsb_release -cs) stable" | sudo tee /etc/apt/sources.list.d/docker.list > /dev/null
+
+apt-get update
+apt-get install -y docker-ce docker-ce-cli containerd.io docker-buildx-plugin docker-compose-plugin
```

Comment:
You might save yourself some time by having an action to create an image with all this already installed. I've user https://www.packer.io/ in the past and it's very easy

Reply:
I'd like to do this in a separate PR. It doesn't take much time to run the user data at start up and it's easier to iterate on.